### PR TITLE
fix: Don't return partial for branded or template keys

### LINKED ISSUE #719: Records with template literal or branded keys return Partial with groupBy
# What

Grouping by branded type and then iterating over entries leads to invalid `undefined` added to the type. 

In Remeda v1 this worked correctly so the behavior has changed.

# Reproduce

```ts
import * as R from 'remeda'

type Branded<K, T> = K & { _type: T }
type UserID = Branded<string, 'UserId'>

const items = [
  { owner: 'U1' as UserID, name: 'Phone' },
  { owner: 'U2' as UserID, name: 'Pencil' },
]

const itemsByOwner = R.groupBy(items, (item) => {
  return item.owner
})

const entries = R.entries(itemsByOwner)

for (const entry of entries) {
  const owner = entry[0]
  const ownersItems = entry[1] // ownersItems can be undefined but it should never be undefined. This seems to be caused by groupBy returning Partial record
  console.log(owner, ownersItems)
}

```

### PR DESCRIPTION
**Pull request by cjquines**
closes #719

### CODE REVIEW

**Comment by cjquines** [comment]
hm why did the codesandbox build fail?

File: src/groupBy.test.ts [hunk 1]
```diff
@@ -176,6 +176,56 @@ describe("typing", () => {
...
+      [
+        { owner: "UserID_1" as UserID, name: "Phone" },
+        { owner: "UserID_2" as UserID, name: "Pencil" },
+      ],
+      (x) => x.owner,
+    );
+    expectTypeOf(data).toEqualTypeOf<
+      Record<UserID, NonEmptyArray<{ owner: UserID; name: string }>>
+    >();
+  });
```
Comment (eranhirsch) [inline src/groupBy.test.ts:227]: I appreciate that testing on the real functions that use the types makes sense, as these are the interfaces our users encounter, but I think it makes sense to narrow the tests to just `IfSimpleRecord` and test it directly with a test file for types.ts... e.g. `internal/types.test.ts`

File: src/groupBy.test.ts [hunk 2]
```diff
@@ -176,6 +176,56 @@ describe("typing", () => {
     >();
   });
 
+  test("symbol", () => {
+    const mySymbol = Symbol("mySymbol");
+    const data = groupBy(
+      [
+        { [mySymbol]: "cat", b: 123 },
+        { [mySymbol]: "dog", b: 456 },
```
Comment (eranhirsch) [inline src/groupBy.test.ts:184]: I'm confused; why does it matter that the object has a prop with `mySymbol`? isn't the grouping function's return value the only thing we care about here?

File: src/groupBy.test.ts [hunk 3]
```diff
@@ -176,6 +176,56 @@ describe("typing", () => {
...
+      ],
+      () => mySymbol,
```
Comment (eranhirsch) [inline src/groupBy.test.ts:186]: use constant?
```suggestion
      constant(mySymbol),
```

File: src/groupBy.test.ts [hunk 4]
```diff
@@ -176,6 +176,56 @@ describe("typing", () => {
...
+  test("branded string", () => {
+    type Branded<K, T> = K & { _type: T };
+    type UserID = Branded<string, "UserId">;
+
+    const data = groupBy(
+      [
+        { owner: "U1" as UserID, name: "Phone" },
+        { owner: "U2" as UserID, name: "Pencil" },
+      ],
+      (x) => x.owner,
```
Comment (eranhirsch) [inline src/groupBy.test.ts:207]: use prop?
```suggestion
      prop("owner"),
```

File: src/groupBy.test.ts [hunk 5]
```diff
@@ -176,6 +176,56 @@ describe("typing", () => {
...
+          typeof mySymbol,
+          NonEmptyArray<{ [mySymbol]: string; b: number }>
+        >
+      >
+    >();
+  });
+
+  test("branded string", () => {
+    type Branded<K, T> = K & { _type: T };
+    type UserID = Branded<string, "UserId">;
```
Comment (eranhirsch) [inline src/groupBy.test.ts:200]: I prefer pulling out types to the module level, I don't have anything smart for a reason, it is just always a bit jarring to see a type definition inside a test

Comment (eranhirsch) [inline src/groupBy.test.ts:200]: We now have an internal branded type, we can use it.

File: src/groupBy.test.ts [hunk 6]
```diff
@@ -176,6 +176,56 @@ describe("typing", () => {
...
+
+    const data = groupBy(
+      [
+        { owner: "U1" as UserID, name: "Phone" },
+        { owner: "U2" as UserID, name: "Pencil" },
+      ],
```
Comment (eranhirsch) [inline src/groupBy.test.ts:206]: you don't need to define items in the array for type tests; it detracts from what the test is about; you can cast an empty array to whatever type you want to test against...

```suggestion
      [] as Array<{ owner: UserId, name: string }>,
```

**Review by eranhirsch — COMMENTED** [review]
This is amazing, I didn't think it will be possible to solve the template string cases.

I want to make sure we cover as many cases of template strings we can think of:
- `${"cat" | "dog"}_${number}`
- `prefix_${string}`
- `prefix_${1 | 2}`
etc...

File: src/internal/types.ts [hunk 7]
```diff
@@ -20,25 +29,44 @@ export type Mapped<T extends IterableContainer, K> = {
...
+ */
+type IsBoundedString<T> = T extends string
+  ? Split<T, "">[number] extends infer U
+    ? [`${number}`] extends [U]
+      ? false
+      : [string] extends [U]
+        ? false
+        : true
+    : never
+  : never;
```
[the remark this answers is not in the document]
Comment (cjquines) [reply src/internal/types.ts:43]: oops this was a mistake, i'll convert to false

File: src/internal/types.ts [hunk 8]
```diff
@@ -20,25 +29,44 @@ export type Mapped<T extends IterableContainer, K> = {
...
+/**
+ * Checks if a type is a bounded key: a union of bounded strings, numeric
+ * literals, or symbol literals.
+ */
+type IsBoundedKey<T> = T extends unknown
+  ? IsStringLiteral<T> extends true
+    ? IsBoundedString<T>
+    : IsNumericLiteral<T> extends true
+      ? true
+      : IsSymbolLiteral<T>
```
[the remark this answers is not in the document]
Comment (cjquines) [reply src/internal/types.ts:54]: hm… in the case that someone does declare a type `Record<symbol, string>`, i think we'd want it to be a simple record, though? like, in theory, someone could write

```ts
const a = {} as Record<symbol, string>;
// etc.
```

though i don't know why they'd do that.

File: src/groupBy.test.ts (hunk 1 above)
Comment (cjquines) [reply src/groupBy.test.ts:227]: agreed, i think i'll just discard these tests

File: src/groupBy.test.ts (hunk 4 above)
Comment (cjquines) [reply src/groupBy.test.ts]: test removed, so this is outdated—i didn't use prop initially for consistency

File: src/groupBy.test.ts (hunk 3 above)
Comment (cjquines) [reply src/groupBy.test.ts]: (test removed)

File: src/groupBy.test.ts (hunk 5 above)
Comment (cjquines) [reply src/groupBy.test.ts]: (test removed)

Comment (cjquines) [reply src/groupBy.test.ts]: right, we do now!

File: src/groupBy.test.ts (hunk 6 above)
Comment (cjquines) [reply src/groupBy.test.ts]: (test removed)

File: src/internal/types.ts [hunk 9]
```diff
@@ -24,25 +33,54 @@ export type Mapped<T extends IterableContainer, K> = {
...
- * combination). Another way of thinking about simple records is that the number
- * of properties are unbound (they can be empty, or they can have 1000 props).
- *
- * TODO: Template string literals (e.g. `prefix_${number}`) should also be
- * considered simple, but we don't know how to check for them...
+ * Check if a type is a simple record. A simple record is one that has an
+ * unbounded number of properties. For example, a record with a union of string
+ * or number keys is simple, and so is a record with `prefix_${number}` keys.
  */
 export type IfSimpleRecord<
```
Comment (eranhirsch) [inline src/internal/types.ts:40]: Now that the difference is the boundness of the key this name isn't very good anymore, it doesn't communicate the difference well. 
```ts
Record<`${MyEnum}_suffix_${number}`, MyObject>
```

isn't any simpler than `{ a: number }`.

File: src/internal/types.test.ts [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    >().toEqualTypeOf<true>();
+
+    expectTypeOf<
+      IfSimpleRecord<Record<number | string, unknown>>
+    >().toEqualTypeOf<true>();
+
+    expectTypeOf<
+      IfSimpleRecord<Record<symbol, unknown>>
+    >().toEqualTypeOf<true>();
+  });
```
Comment (eranhirsch) [inline src/internal/types.test.ts:20]: no reason not to split them up, and maybe have a slightly more descriptive name 🙄 ?

**Review by eranhirsch — COMMENTED** [review]
This is really impressive! Good job!

File: src/internal/types.test.ts [hunk 11]
```diff
@@ -0,0 +1,99 @@
+import { type Branded, type IfSimpleRecord } from "./types";
+
+describe("IfSimpleRecord", () => {
```
[the remark this answers is not in the document]
Comment (cjquines) [reply src/internal/types.test.ts:3]: 1 should already be covered? `Record<a, string>` is the same as `{ a: string }`

but good point with 2, i added some tests

File: src/internal/types.ts (hunk 9 above)
Comment (cjquines) [reply src/internal/types.ts:40]: renamed to `IfBoundedRecord` (and moved around the args)

File: src/internal/types.ts [hunk 12]
```diff
@@ -24,25 +34,57 @@ export type Mapped<T extends IterableContainer, K> = {
 export type IterableContainer<T = unknown> = ReadonlyArray<T> | readonly [];
 
 /**
- * We define a Simple record as one that doesn't define it's properties beyond
- * their general type. This is the case for string and number keys (and their
- * combination). Another way of thinking about simple records is that the number
- * of properties are unbound (they can be empty, or they can have 1000 props).
+ * Check if a type is guaranteed to be a bounded record: a record with a fixed
```
Comment (eranhirsch) [inline src/internal/types.ts:37]: finite?

```suggestion
 * Check if a type is guaranteed to be a bounded record: a record with a finite
```

File: src/internal/types.ts [hunk 13]
```diff
@@ -24,25 +34,57 @@ export type Mapped<T extends IterableContainer, K> = {
...
+ * set of keys.
  *
- * TODO: Template string literals (e.g. `prefix_${number}`) should also be
- * considered simple, but we don't know how to check for them...
+ * For example, a record with a union of string or number keys is unbounded,
```
Comment (eranhirsch) [inline src/internal/types.ts:40]: 1. Would this be better as a JSDoc `@example`? 
2. Would this be clearer in code vs words (e.g. `Record<string, unknown>`)
3. Why start with a complex example (union) instead of the trivial one?

**Review by eranhirsch — APPROVED** [review]
